File: sdk/nova_sso.py
  1→"""
  2→NovaSSO Python SDK
  3→业务系统接入统一登录的客户端库
  4→
  5→使用方式:
  6→    from nova_sso import NovaSSOClient
  7→
  8→    sso = NovaSSOClient(
  9→        sso_url="https://sso.yourdomain.dn42",
 10→        app_id="your-app-id",
 11→        callback_url="https://your-app/sso/callback"
 12→    )
 13→
 14→    # Flask 示例
 15→    @app.route("/login")
 16→    def login():
 17→        return redirect(sso.get_login_url())
 18→
 19→    @app.route("/sso/callback")
 20→    def sso_callback():
 21→        ticket = request.args.get("ticket")
 22→        user = sso.validate_ticket(ticket)
 23→        if user:
 24→            session["user"] = user
 25→            return redirect("/")
 26→        return "登录失败"
 27→"""
 28→
 29→import requests
 30→from urllib.parse import urlencode, urlparse
 31→from typing import Optional, Dict
 32→
 33→
 34→class NovaSSOClient:
 35→    """NovaSSO 客户端"""
 36→
 37→    def __init__(self, sso_url: str, app_id: str, callback_url: str,
 38→                 app_secret: str = None, timeout: int = 10):
 39→        """
 40→        初始化 SSO 客户端
 41→
 42→        Args:
 43→            sso_url: SSO 服务器地址，如 https://sso.example.com
 44→            app_id: 应用ID（在SSO管理后台创建）
 45→            callback_url: 回调地址（必须与SSO中配置的一致）
 46→            app_secret: 应用密钥（可选，用于增强验证）
 47→            timeout: 请求超时时间（秒）
 48→        """
 49→        self.sso_url = sso_url.rstrip("/")
 50→        self.app_id = app_id
 51→        self.callback_url = callback_url
 52→        self.app_secret = app_secret
 53→        self.timeout = timeout
 54→
 55→    def get_login_url(self, redirect: str = "", silent: bool = False) -> str:
 56→        """
 57→        获取登录跳转URL
 58→
 59→        Args:
 60→            redirect: 登录成功后跳转的地址（可选）
 61→            silent: 是否静默模式（无感登录用）
 62→
 63→        Returns:
 64→            登录页面URL
 65→        """
 66→        params = {
 67→            "service": self.callback_url,
 68→            "app": self.app_id,
 69→        }
 70→        if redirect:
 71→            params["redirect"] = redirect
 72→        if silent:
 73→            params["silent"] = "true"
 74→
 75→        return f"{self.sso_url}/login?{urlencode(params)}"
 76→
 77→    def get_logout_url(self, redirect: str = "") -> str:
 78→        """
 79→        获取登出URL
 80→
 81→        Args:
 82→            redirect: 登出后跳转地址
 83→
 84→        Returns:
 85→            登出URL
 86→        """
 87→        params = {}
 88→        if redirect:
 89→            params["redirect"] = redirect
 90→        url = f"{self.sso_url}/logout"
 91→        if params:
 92→            url += f"?{urlencode(params)}"
 93→        return url
 94→
 95→    def validate_ticket(self, ticket: str) -> Optional[Dict]:
 96→        """
 97→        验证服务票据（ST）
 98→
 99→        Args:
            ticket: 服务票据（从回调URL的ticket参数获取）

        Returns:
            用户信息字典，验证失败返回None
            {
                "user_id": 1,
                "username": "xxx",
                "email": "xxx@xx.com",
                "nickname": "xxx",
                "avatar": "...",
                "is_admin": false,
                "tgt_id": "TGT-xxx",
                "app_id": "xxx"
            }
        """
        if not ticket or not ticket.startswith("ST-"):
            return None

        try:
            resp = requests.get(
                f"{self.sso_url}/api/serviceValidate",
                params={
                    "ticket": ticket,
                    "service": self.callback_url,
                    "app": self.app_id,
                },
                timeout=self.timeout
            )
            data = resp.json()

            if data.get("success"):
                return data.get("user")
            return None

        except requests.RequestException as e:
            print(f"[NovaSSO] 验证票据失败: {e}")
            return None

    def get_user_info(self, user_id: int) -> Optional[Dict]:
        """获取用户信息（需管理员权限，暂不实现）"""
        # 预留接口，后续可扩展
        return None

    # ===== Flask 便捷装饰器 =====

    def flask_login_required(self, f):
        """
        Flask 登录验证装饰器

        使用方式:
            @app.route("/protected")
            @sso.flask_login_required
            def protected():
                return f"Hello, {session['user']['username']}"
        """
        from functools import wraps
        from flask import session, redirect, request, url_for

        @wraps(f)
        def decorated(*args, **kwargs):
            if "user" in session:
                return f(*args, **kwargs)

            # 检查是否有ST回调
            ticket = request.args.get("ticket")
            if ticket:
                user = self.validate_ticket(ticket)
                if user:
                    session["user"] = user
                    # 重定向到干净的URL（去掉ticket参数）
                    return redirect(request.path)

            # 跳转到SSO登录
            return redirect(self.get_login_url(redirect=request.url))

        return decorated

    # ===== FastAPI 便捷依赖 =====

    def fastapi_get_current_user(self):
        """
        FastAPI 依赖注入函数

        使用方式:
            from fastapi import Depends, HTTPException

            @app.get("/protected")
            async def protected(user = Depends(sso.fastapi_get_current_user)):
                return {"user": user}
        """
        from fastapi import Request, HTTPException, Cookie
        from typing import Optional

        async def _get_user(request: Request, ticket: Optional[str] = None):
            # 1. 从session中获取
            user = request.session.get("user") if hasattr(request, 'session') else None
            if user:
                return user

            # 2. 验证ticket
            if ticket:
                user_info = self.validate_ticket(ticket)
                if user_info:
                    if hasattr(request, 'session'):
                        request.session["user"] = user_info
                    return user_info

            raise HTTPException(
                status_code=302,
                headers={"Location": self.get_login_url(redirect=str(request.url))}
            )

        return _get_user


# ===== 便捷函数 =====

def create_sso_client(sso_url: str, app_id: str, callback_url: str,
                      **kwargs) -> NovaSSOClient:
    """便捷创建SSO客户端"""
    return NovaSSOClient(sso_url, app_id, callback_url, **kwargs)

File: sdk/nova-sso.js
  1→/**
  2→ * NovaSSO 前端无感登录 SDK
  3→ * 基于隐藏 iframe + postMessage 实现跨域无感登录
  4→ *
  5→ * 使用方式:
  6→ *   const sso = new NovaSSO({
  7→ *     ssoUrl: 'https://sso.yourdomain.dn42',
  8→ *     callbackUrl: 'https://your-app/sso/silent-callback.html'
  9→ *   });
 10→ *
 11→ *   // 页面加载时尝试无感登录
 12→ *   sso.trySilentLogin().then(user => {
 13→ *     console.log('登录成功', user);
 14→ *   }).catch(() => {
 15→ *     console.log('需要手动登录');
 16→ *   });
 17→ */
 18→
 19→class NovaSSO {
 20→  constructor(options) {
 21→    this.ssoUrl = options.ssoUrl.replace(/\/$/, '');
 22→    this.callbackUrl = options.callbackUrl;
 23→    this.appId = options.appId || '';
 24→    this.timeout = options.timeout || 5000;
 25→    this.user = null;
 26→    this._storageKey = 'nova_sso_user';
 27→
 28→    // 从本地存储恢复用户信息
 29→    const saved = localStorage.getItem(this._storageKey);
 30→    if (saved) {
 31→      try {
 32→        this.user = JSON.parse(saved);
 33→      } catch (e) {}
 34→    }
 35→  }
 36→
 37→  /**
 38→   * 尝试无感登录
 39→   * @returns {Promise<Object>} 用户信息
 40→   */
 41→  async trySilentLogin() {
 42→    return new Promise((resolve, reject) => {
 43→      // 已有用户信息，直接返回
 44→      if (this.user) {
 45→        resolve(this.user);
 46→        return;
 47→      }
 48→
 49→      const iframe = document.createElement('iframe');
 50→      iframe.style.display = 'none';
 51→      iframe.style.position = 'absolute';
 52→      iframe.style.width = '0';
 53→      iframe.style.height = '0';
 54→      iframe.style.border = 'none';
 55→      iframe.setAttribute('sandbox', 'allow-scripts allow-same-origin allow-forms');
 56→
 57→      const params = new URLSearchParams({
 58→        service: this.callbackUrl,
 59→        app: this.appId,
 60→        silent: 'true'
 61→      });
 62→
 63→      iframe.src = `${this.ssoUrl}/login?${params.toString()}`;
 64→
 65→      let resolved = false;
 66→
 67→      const messageHandler = (e) => {
 68→        // 验证消息来源（生产环境请精确匹配）
 69→        if (e.data && e.data.type === 'nova_sso_auth') {
 70→          resolved = true;
 71→          this._cleanup(iframe, messageHandler, timeoutId);
 72→
 73→          if (e.data.success && e.data.user) {
 74→            this.user = e.data.user;
 75→            localStorage.setItem(this._storageKey, JSON.stringify(this.user));
 76→            resolve(this.user);
 77→          } else {
 78→            reject(new Error(e.data.error || '登录失败'));
 79→          }
 80→        }
 81→      };
 82→
 83→      const timeoutId = setTimeout(() => {
 84→        if (!resolved) {
 85→          this._cleanup(iframe, messageHandler, timeoutId);
 86→          reject(new Error('SSO 登录超时'));
 87→        }
 88→      }, this.timeout);
 89→
 90→      window.addEventListener('message', messageHandler);
 91→      document.body.appendChild(iframe);
 92→    });
 93→  }
 94→
 95→  /**
 96→   * 跳转到登录页
 97→   * @param {string} redirect 登录后跳转地址
 98→   */
 99→  redirectToLogin(redirect = '') {
    const params = new URLSearchParams({
      service: this.callbackUrl,
      app: this.appId
    });
    if (redirect) {
      params.set('redirect', redirect);
    }
    window.location.href = `${this.ssoUrl}/login?${params.toString()}`;
  }

  /**
   * 跳转到登出页
   * @param {string} redirect 登出后跳转地址
   */
  redirectToLogout(redirect = '') {
    this.user = null;
    localStorage.removeItem(this._storageKey);
    const params = redirect ? `?redirect=${encodeURIComponent(redirect)}` : '';
    window.location.href = `${this.ssoUrl}/logout${params}`;
  }

  /**
   * 获取当前用户
   * @returns {Object|null}
   */
  getCurrentUser() {
    return this.user;
  }

  /**
   * 检查是否已登录
   * @returns {boolean}
   */
  isLoggedIn() {
    return !!this.user;
  }

  /**
   * 清除本地登录状态
   */
  clearLocalAuth() {
    this.user = null;
    localStorage.removeItem(this._storageKey);
  }

  // 清理
  _cleanup(iframe, handler, timeoutId) {
    window.removeEventListener('message', handler);
    if (timeoutId) clearTimeout(timeoutId);
    if (iframe.parentNode) {
      iframe.parentNode.removeChild(iframe);
    }
  }
}

/**
 * 静默回调页脚本
 * 在 silent-callback.html 中引入
 * 作用：接收ST，验证后通过 postMessage 通知父页面
 */
NovaSSO.handleSilentCallback = async function(validateUrl = '/api/sso-validate') {
  const params = new URLSearchParams(window.location.search);
  const ticket = params.get('ticket');
  const error = params.get('error');

  const sendMessage = (data) => {
    if (window.parent && window.parent !== window) {
      window.parent.postMessage({
        type: 'nova_sso_auth',
        ...data
      }, '*'); // 生产环境请替换为具体的 origin
    }
  };

  if (ticket) {
    try {
      const resp = await fetch(`${validateUrl}?ticket=${encodeURIComponent(ticket)}`);
      const data = await resp.json();
      if (data.success) {
        sendMessage({ success: true, user: data.user });
      } else {
        sendMessage({ success: false, error: data.error || '验证失败' });
      }
    } catch (e) {
      sendMessage({ success: false, error: e.message });
    }
  } else {
    sendMessage({ success: false, error: error || '需要登录' });
  }
};

// CommonJS / ES Module 兼容
if (typeof module !== 'undefined' && module.exports) {
  module.exports = NovaSSO;
}
if (typeof window !== 'undefined') {
  window.NovaSSO = NovaSSO;
}